src/omega_edit/client.ts

```diff
@@ -93,56 +93,52 @@ export function activate(ctx: vscode.ExtensionContext) {
         filePath: fileToEdit,
       })
 
+      let fileSize = await omegaEditSession.getComputedFileSize(s)
+
       let vpAll = await omegaEditViewport.createViewport(
-        randomId().toString(),
-        s,
-        0,
-        1000
-      )
-      let vp1 = await omegaEditViewport.createViewport(
-        randomId().toString(),
+        '',
         s,
         0,
-        64
-      )
-      let vp2 = await omegaEditViewport.createViewport(
-        randomId().toString(),
-        s,
-        64,
-        64
-      )
-      let vp3 = await omegaEditViewport.createViewport(
-        randomId().toString(),
-        s,
-        128,
-        64
+        fileSize,
+        false
       )
 
+      // let data = await omegaEditViewport.getViewportData(vpAll)
+      let vp1 = await omegaEditViewport.createViewport('', s, 0, 64, false)
+      let vp2 = await omegaEditViewport.createViewport('', s, 64, 64, false)
+      let vp3 = await omegaEditViewport.createViewport('', s, 128, 64, false)
+
       await viewportSubscribe(panel, vpAll, vpAll, 'vpAll', 'hexAll')
       await viewportSubscribe(panel, vpAll, vp1, 'viewport1', null)
       await viewportSubscribe(panel, vpAll, vp2, 'viewport2', null)
       await viewportSubscribe(panel, vpAll, vp3, 'viewport3', null)
 
       panel.webview.onDidReceiveMessage(
         async (message) => {
-          if (message.command === 'printChangeCount') {
-            vscode.window.showInformationMessage(message.changeCount)
-            return
-          }
           var omegaEdit = new OmegaEdit(
             s,
             message.offset,
-            message.command === 'overwriteByte'
-              ? `${message.deleteValue},${message.addValue}`
-              : message.data,
+            message.data,
             message.len,
             panel
           )
+
+          var fileSize = await omegaEditSession.getComputedFileSize(s)
+          var searchPattern = message.searchPattern ? message.searchPattern : ''
+
+          if (searchPattern !== '' && fileSize < searchPattern.length) {
+            throw new Error("Search pattern can't be larger than file")
+          }
```

Comment:
So data could be a small number of bytes, but the search pattern might be using hex, or even individual bits (someday) so this notion that something is wrong if the pattern is larger than the data may not hold. For now it seems sensible, but I can see this having to be taken out in the future. 

E.g., There's a complete positive VMF test message that is 11 bytes long, and given the bit-packed nature of it, searching in it for a bit string might use things like search for "0110001" (0x31, but as 7 bits only), in which case the patterns, could be, in bytes (0 or 1 are each a byte), bigger than the data in bytes.

Reply:
So this check is for making sure when a pattern is given that it is not bigger than the size of file being searched.

Reply:
> So this check is for making sure when a pattern is given that it is not bigger than the size of file being searched.

Understood, but why does that matter? Do things break if the pattern is larger than the input data?

Reply:
Yeah I believe it causes the server to crash but @scholarsmate would be better at explaining why than myself

Reply:
Seems like a bug somewhere else?

Reply:
If the needle is larger than the haystack, the library will return a nullptr, indicating a search error, rather than a lack of matches.  The check in question happens here: https://github.com/ctc-oss/omega-edit/blob/main/src/lib/search.cpp#L47.  I'm open to discussion on how and where we think such problems should be handled, but I'd prefer the bad input not to be sent in the first place, thus this sanity check being done early in the client.  Other options include ignoring the problem and returning an empty list of matches, or having to bubble up error status and dealing with it after a round trip has been taken.

Reply:
Hex, Octal, ASCII, etc. are presentation details (views) of the same underlying bit pattern (model).  At this level in the stack, we're sending the bit pattern (in bytes), so it is expected to always be no greater than the haystack.  The smallest addressable unit is a byte.

If we want to search for bit patterns that are not aligned on byte boundaries then we can use bit shifts and if the bit patterns are not divisible by 8, then we can use masks.  This advanced searching feature (or corresponding
replace) hasn't been implemented in Ωedit yet.  Currently it only works on aligned bytes.

Reply:
>  but I'd prefer the bad input not to be sent in the first place,

I'm not sure I'd consider that bad input, I'd just consider it a search pattern that just will never match. If you do `myString.indexOf("foo")`, java doesn't throw an exception if "foo" is larger than the length of `myString`. It's not an error, it just won't match and returns the "no match" value of -1.

>  I'm open to discussion on how and where we think such problems should be handled,

IMO, I wouldn't treat it any different than any other search pattern. Send the pattern to whatever does the search, return no matches if the pattern doesn't match (and it won't in the case of too large data), and handle no match like any other pattern that has no matches.


That said, it's fine with me if that's the design decision, but I think if this check is going to be made here it needs a comment as to why it's there, because it feels non-standard to me. Especially if it could lead to a null pointer exception if someone removed it because they decided they didn't want what appears to just be an optimization.

Reply:
Okay, I'm good with the comment approach.

Since we have the data right at our fingertips, and a simple check tells us that no matches are possible, I don't see a need to take the trip even if I modified the Ωedit code to return an empty set.  If I change the behavior in Ωedit later, this code is still fine, requiring no changes.

Reply:
@Shanedell, here is the comment I'd propose adding above this check:

          // If the search pattern exceeds the length of the file, matches are
          // not possible.  Ωedit (as implemented currently) considers
          // patterns that are longer than the length of the file to be an
          // error (it will return a null pointer instead of an empty list).